### PR TITLE
Bigint divide

### PR DESCRIPTION
**Pull request by bretambrose**
Division and modulus support

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/bigint.c [hunk 1]
```diff
@@ -958,3 +1048,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+    struct aws_bigint *quotient,
+    struct aws_bigint *remainder,
+    const struct aws_bigint *dividend,
+    uint32_t divisor) {
+
+    AWS_PRECONDITION(aws_bigint_is_valid(quotient));
+    AWS_PRECONDITION(aws_bigint_is_valid(remainder));
+    AWS_PRECONDITION(aws_bigint_is_valid(dividend));
+
+    size_t quotient_length = aws_array_list_length(&dividend->digits);
```
Comment (justinboswell) [inline source/bigint.c]: Nitpick: const?

EDIT: After reading this a few times, I think const would really help me visualize the moving parts here. I think a quick const pass would be a big help.

File: source/bigint.c [hunk 2]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+
+    struct aws_bigint *temp_quotient = s_aws_bigint_new_reserved(quotient->digits.alloc, quotient_length);
+    if (temp_quotient == NULL) {
+        return AWS_OP_ERR;
+    }
+
+    uint64_t wide_divisor = divisor;
```
Comment (justinboswell) [inline source/bigint.c]: const

File: source/bigint.c [hunk 3]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+
+        current_remainder = two_digit_dividend % wide_divisor;
+    }
+
+    s_aws_bigint_trim_leading_zeros(temp_quotient);
+
+    aws_array_list_swap_contents(&temp_quotient->digits, &quotient->digits);
+    quotient->sign = 1;
+
+    uint32_t final_remainder = (uint32_t)current_remainder;
```
Comment (justinboswell) [inline source/bigint.c]: const

File: source/bigint.c [hunk 4]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+
+    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
+    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+
+    size_t lhs_digit_count = aws_array_list_length(&lhs->digits);
+    size_t rhs_digit_count = aws_array_list_length(&rhs->digits);
+    AWS_FATAL_ASSERT(lhs_digit_count > rhs_digit_count); /* padding by zero */
+    AWS_FATAL_ASSERT(rhs_digit_count >= 2);
+
+    size_t quotient_digit_count = lhs_digit_count - rhs_digit_count;
```
Comment (justinboswell) [inline source/bigint.c]: const

File: source/bigint.c [hunk 5]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+static int s_aws_bigint_normalized_divide(
+    struct aws_bigint **quotient,
+    struct aws_bigint **remainder,
+    struct aws_bigint *lhs,
+    const struct aws_bigint *rhs) {
+
+    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
+    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+
+    size_t lhs_digit_count = aws_array_list_length(&lhs->digits);
```
Comment (justinboswell) [inline source/bigint.c]: const both

File: source/bigint.c [hunk 6]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+
+    AWS_POSTCONDITION(aws_bigint_is_valid(bigint));
+
+    return shift_count;
+}
+
+/*
+ * General algorithm requirements/assumptions:
+ *   (1) lhs >= rhs
+ *   (2) lhs, rhs both positive
```
Comment (justinboswell) [inline source/bigint.c]: Should we assert on these?

File: source/bigint.c [hunk 7]
```diff
@@ -961,3 +1051,370 @@ int aws_bigint_shift_left(struct aws_bigint *bigint, size_t shift_amount) {
...
+    struct aws_bigint **remainder,
+    struct aws_bigint *lhs,
+    const struct aws_bigint *rhs) {
+
+    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
+    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+
+    size_t lhs_digit_count = aws_array_list_length(&lhs->digits);
+    size_t rhs_digit_count = aws_array_list_length(&rhs->digits);
+    AWS_FATAL_ASSERT(lhs_digit_count > rhs_digit_count); /* padding by zero */
```
Comment (justinboswell) [inline source/bigint.c]: Couldn't they also be equal, as long as the most significant digit is lhs > rhs?

Comment (bretambrose) [reply source/bigint.c]: We pad with a leading zero to prevent that.

File: source/bigint.c (hunk 6 above)
Comment (bretambrose) [reply source/bigint.c]: Added

**Pull request merged**